intellij-plugin-structure/structure-intellij/src/main/java/com/jetbrains/plugin/structure/intellij/verifiers/PluginUntilBuildVerifier.kt

```diff
@@ -0,0 +1,100 @@
+package com.jetbrains.plugin.structure.intellij.verifiers
+
+import com.jetbrains.plugin.structure.intellij.beans.PluginBean
+import com.jetbrains.plugin.structure.intellij.problems.InvalidUntilBuild
+import com.jetbrains.plugin.structure.intellij.problems.InvalidUntilBuildWithJustBranch
+import com.jetbrains.plugin.structure.intellij.problems.NonexistentReleaseInUntilBuild
+import com.jetbrains.plugin.structure.intellij.problems.ProductCodePrefixInBuild
+import com.jetbrains.plugin.structure.intellij.problems.SuspiciousUntilBuild
+import com.jetbrains.plugin.structure.intellij.version.IdeVersion
+
+
+private const val BUILD_NUMBER = "__BUILD_NUMBER__"
+private const val SNAPSHOT = "SNAPSHOT"
+
+private const val SUSPICIOUS_BASELINE_LOWER_BOUND = 281
+private const val FIRST_YEARLY_BASED_RELEASE_NUMBER_BASELINE = 162
+
+class PluginUntilBuildVerifier {
+  fun verify(plugin: PluginBean,
+             descriptorPath: String,
+             problemRegistrar: ProblemRegistrar) = with(problemRegistrar) {
+    val untilBuild = plugin.ideaVersion?.untilBuild ?: return
+    if (isJustASingleComponent(untilBuild)) {
+      if (isSpecialSingleComponent(untilBuild)) {
+        return
+      }
+      registerProblem(InvalidUntilBuildWithJustBranch(descriptorPath, untilBuild))
+      verifySingleComponentUntilBuild(untilBuild, descriptorPath, problemRegistrar)
+      return
+    }
+
+    val untilBuildParsed = IdeVersion.createIdeVersionIfValid(untilBuild)
+    if (untilBuildParsed == null) {
+      registerProblem(InvalidUntilBuild(descriptorPath, untilBuild))
+    } else {
+      verifyBaseLineVersion(untilBuildParsed, untilBuild, descriptorPath, problemRegistrar)
+      if (untilBuildParsed.productCode.isNotEmpty()) {
+        registerProblem(ProductCodePrefixInBuild(descriptorPath))
+      }
+    }
+  }
+
+  private fun verifyBaseLineVersion(untilBuild: IdeVersion,
+                                    untilBuildValue: String,
+                                    descriptorPath: String,
+                                    problemRegistrar: ProblemRegistrar) = with(problemRegistrar) {
+    untilBuild.baselineVersion.let {
+      verifyBaseline(it, untilBuildValue, untilBuild, descriptorPath, problemRegistrar)
+    }
+  }
+
+  private fun verifyBaseline(baseline: Int,
+                             untilBuildValue: String,
+                             untilBuild: IdeVersion?,
+                             descriptorPath: String,
+                             problemRegistrar: ProblemRegistrar) = with(problemRegistrar) {
+    if (baseline >= 999) {
+      registerProblem(InvalidUntilBuild(descriptorPath, untilBuildValue, untilBuild))
```

Comment:
Since we have a lot of existing plugins with 999 in until, I see the following problem: when PV tries to verify the plugin with such until (or plugins which have dependencies on such plugin), it will get the PluginCreationFail result from structure, and won't be able to make further verifications. Or I'm not right?

Suppose, that `unacceptable warning` should help here.

Reply:
Updated to `unnaceptable warnings`.